onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -339,55 +509,84 @@ Status Attention<T>::Compute(OpKernelContext* context) const {
         const int qkv_index = static_cast<int>(i % 3);
 
         int input_offset = batch_index * sequence_length * input_hidden_size;
-        int weights_offset = qkv_index * hidden_size + head_index * head_size;
+
         T* qkv_dest = QKV[qkv_index];
-        int qkv_offset = (batch_index * num_heads_ + head_index) * (sequence_length * head_size);
+        int qkv_offset = 0;
+        int head_size_passed_in = 0;
+
+        if (qkv_index == 0) {
+          head_size_passed_in = q_head_size;
+        } else if (qkv_index == 1) {
+          head_size_passed_in = k_head_size;
+        } else {
+          head_size_passed_in = v_head_size;
+        }
+
+        int weights_offset = 0;
+        int bias_offset = qkv_index * q_hidden_size + head_index * head_size_passed_in;
+
+        if (q_packed_weights_ == nullptr) {
+          weights_offset = bias_offset;
+        } else {
+          weights_offset = head_index * head_size_passed_in;
+        }
+
+        qkv_offset += (batch_index * num_heads_ + head_index) * (sequence_length * head_size_passed_in);
 
         // TODO!! memcpy here makes it not worthwhile to use Gemm batch. Possible to post process?
-        // broadcast 3NH -> (3.B.N.S.H)
-        const T* broadcast_data_src = bias_data + weights_offset;
+        // broadcast NH -> (B.N.S.H) for each of Q, K, V
+
+        const T* broadcast_data_src = bias_data + bias_offset;
         T* broadcast_data_dest = QKV[qkv_index] + qkv_offset;
+
         for (int seq_index = 0; seq_index < sequence_length; seq_index++) {
-          memcpy(broadcast_data_dest, broadcast_data_src, head_size * sizeof(T));
-          broadcast_data_dest += head_size;
+          memcpy(broadcast_data_dest, broadcast_data_src, head_size_passed_in * sizeof(T));
+          broadcast_data_dest += head_size_passed_in;
         }
 
         //                   original           transposed            iteration
         // A: input          (BxSxD)            (B.)S x D             S x D
-        // B: weights        (Dx3xNxH)          D x (3.N.)H           D x H
-        // C: QKV[qkv_index] (3xBxNxSxH)        (3.B.N.)S x H         S x H
```

Comment:
It might help to leave the comments for B and C (add comments that they are for case that q,k,v hidden size is same)